### ISSUE TITLE
CQEngine can provide an expiration strategy to prevent the memory space from being used up

### ISSUE THREAD

**Issue by clncon** [npgall/cqengine#287]
I hope that CQEngine can provide some expiration strategies to prevent the collection from growing

**Comment by shenzhaoquan** [comment]
I also encountered this problem and added a simple implementation CapacityLimitedIndexedCollection

**Comment by npgall** [comment]
The best way to handle that is probably to add a timestamp in your objects
which indicates the time it was added to the collection. Then define an
attribute which reads that timestamp. Then add an index on that attribute.

Then, you can write a query which retrieves objects which were added a long
time ago, and remove those from the collection.

**Comment by boneylw** [comment]
> The best way to handle that is probably to add a timestamp in your objects which indicates the time it was added to the collection. Then define an attribute which reads that timestamp. Then add an index on that attribute. Then, you can write a query which retrieves objects which were added a long time ago, and remove those from the collection.
> […](#)

I don't think this is a good way, because we can't know when to clear them. It is possible that OOM occurred before I triggered the cleanup task.

In some cases, users cannot accurately assess how much data is entered into the CQEngine collection. This can easily cause memory overflow problems. A collection with limited capacity can better protect applications using CQEngine.

**Comment by npgall** [comment]
CQEngine is for running queries on a dataset. If objects are arbitrarily
removed from the dataset, then queries will return incorrect results, right?

I fail to see how what you propose could be implemented without sacrificing
query correctness.

OTOH, the application might know which objects can be removed safely
without compromising correctness for the particular domain of the
application. So it seems sensible to me to have the application manage that
itself.
